tpm2_pytss/types.py

```diff
@@ -921,54 +921,55 @@ def Unmarshal(cls, buf):
 
 
 class TPM2B_OBJECT(TPM_OBJECT):
+    @classmethod
+    def _get_bytefield(cls):
+        tipe = ffi.typeof(f"{cls.__name__}")
+        for f in tipe.fields:
+            if f[0] != "size":
+                return f[0]
+
     def __init__(self, _cdata=None, **kwargs):
         _cdata, kwargs = fixup_cdata_kwargs(self, _cdata, kwargs)
         super().__init__(_cdata, **kwargs)
 
-        # we don't set size on nested compount TPM2B structures, like TPM2B_PUBLIC
-        tipe = ffi.typeof(_cdata)
-        if "TPM2B_DIGEST" not in tipe.cname:
-            return
-
-        # if their is nothing to update OR the caller passed both the
-        # buffer AND the size, their is nothing to do.
-        if len(kwargs) == 0 or len(kwargs) == 2:
-            return
-
-        key = [*kwargs][0]
-        value = kwargs[key]
-        if len(kwargs) == 0:
-            return
-
-        # update the len
-        _cdata.size = len(value)
+    def __setattr__(self, key, value):
+        _bytefield = type(self)._get_bytefield()
+        if key == "size":
+            raise AttributeError(f"{key} is read only")
+        super().__setattr__(key, value)
+        if key == _bytefield:
+            self._cdata.size = len(value)
```

Comment:
LGTM now complains in the __init__ for TPM_OBJECT that:
`NEWCall to self.__setattr__ in __init__ method, which is overridden by method TPM2B_OBJECT.__setattr__.`

We should probably have the caller in TPM_OBJECT call:
_cdata.__setattr_(k, v) to avoid that.

Reply:
Using `_cdata.__setattr__(k, v)` in TPM_OBJECTs `__init__` breaks some functionality, then you need to pass cdata for everything

Reply:
> Using `_cdata.__setattr__(k, v)` in TPM_OBJECTs `__init__` breaks some functionality, then you need to pass cdata for everything

Ahh Ok, according to https://lgtm.com/rules/10030085/, we want the base class fully initialized before calling to things that are overridden by the subclass. Perhaps we want TPM_OBJECT.__setattr__(self, k, v), so we specifically call the base class setattr here?

Reply:
I wonder something like `super().__init__(_cdata)` and then loop over kwargs and running `super().__setattr__(k, v)` in TPM2B_SIMPLE_OBJECT might work instead

Reply:
test_TPM2B_AUTH_bad_fields fails then

Reply:
Found an ugly hack, lets see if LGTM will accept it

Reply:
> Found an ugly hack, lets see if LGTM will accept it

Yeah I thought multiple init calls would trigger a warning, yep:
```
1 for Multiple calls to `__init__` during object initialization
1 for Unused import
1 for `__init__` method calls overridden method
```

On the plus side we got a different result :-p

Reply:
> test_TPM2B_AUTH_bad_fields fails then

Is that the only test that fails? Maybe some logic change to the base class? It must be calling setattr on the object over cdata which will work. Perhaps in the constructor we limit the values that can be setattr'd to valid fields in the cdata?

Reply:
yeah, exactly, the field check is in the init method, so perhaps a `_setfield` method in TPM_OBJECT that only handles cdata fields and use that in its init method and  `__setattr__`

Reply:
yeah, the fields check is in `__init__`,